<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo">
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


***Note***:- If you are working in IBM Cloud Watson Studio, please replace the command for installing nbformat from `!pip install nbformat==4.2.0` to simply `!pip install nbformat`


In [238]:
!pip install yfinance==0.1.67
!mamba install bs4==4.10.0 -y
!pip install nbformat








  Using cached yfinance-0.1.67-py2.py3-none-any.whl (25 kB)
  Attempting uninstall: yfinance
    Found existing installation: yfinance 0.2.4
    Uninstalling yfinance-0.2.4:
      Successfully uninstalled yfinance-0.2.4


You should consider upgrading via the 'C:\Users\ThakurRa2\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.
'mamba' is not recognized as an internal or external command,
operable program or batch file.


You should consider upgrading via the 'C:\Users\ThakurRa2\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [4]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup


In [59]:
!pip install --upgrade pandas
!pip install --upgrade yfinance

You should consider upgrading via the 'C:\Users\ThakurRa2\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


     ---------------------------------------- 72.4/72.4 KB ? eta 0:00:00
     ---------------------------------------- 2.9/2.9 MB 17.1 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
  Using cached requests-2.31.0-py3-none-any.whl (62 kB)
  Using cached lxml-5.1.0-cp310-cp310-win_amd64.whl (3.9 MB)
  Created wheel for peewee: filename=peewee-3.17.0-py3-none-any.whl size=135766 sha256=c6f597c3359cfd5107c7945834cc15a6d24f641f339564472b1b74b344e75796
  Stored in directory: c:\users\thakurra2\appdata\local\pip\cache\wheels\c7\70\ad\212867e96e7004265a69c4aa5dcff00a95f547a67ba26e7e76
Successfully built peewee
  Attempting uninstall: requests
    Found existing installation: requests 2.26.0
    Unin

You should consider upgrading via the 'C:\Users\ThakurRa2\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.


In [5]:
import plotly.graph_objects as go
from plotly.subplots import make_subplots

In Python, you can ignore warnings using the warnings module. You can use the filterwarnings function to filter or ignore specific warning messages or categories.


In [6]:
import warnings
# Ignore all warnings
warnings.filterwarnings("ignore", category=FutureWarning)

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [7]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [8]:
tesla = yf.Ticker("TSLA")



In [9]:
!wget https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/data/tesla.json

'wget' is not recognized as an internal or external command,
operable program or batch file.


Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [10]:
tesla_data = tesla.history(period="max")


**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [190]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

level_0  index                      Date      Open      High       Low  \
0        0      0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333   
1        1      1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333   
2        2      2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333   
3        3      3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333   
4        4      4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333   

      Close     Volume  Dividends  Stock Splits  
0  1.592667  281494500        0.0           0.0  
1  1.588667  257806500        0.0           0.0  
2  1.464000  123282000        0.0           0.0  
3  1.280000   77097000        0.0           0.0  
4  1.074000  103003500        0.0           0.0

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [137]:
url ="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm"
html_data=requests.get(url).text

Parse the html data using `beautiful_soup`.


In [144]:
soup = BeautifulSoup(html_data, "html.parser")


In [150]:
for index,table in enumerate(tables):
    if ("Tesla Annual Revenue" in str(table)):
        table_index = index
print(table_index)
print(tables[table_index].prettify())

0
<table class="historical_data_table table">
 <thead>
  <tr>
   <th colspan="2" style="text-align:center">
    Tesla Annual Revenue
    <br/>
    <span style="font-size:14px;">
     (Millions of US $)
    </span>
   </th>
  </tr>
 </thead>
 <tbody>
  <tr>
   <td style="text-align:center">
    2021
   </td>
   <td style="text-align:center">
    $53,823
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2020
   </td>
   <td style="text-align:center">
    $31,536
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2019
   </td>
   <td style="text-align:center">
    $24,578
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2018
   </td>
   <td style="text-align:center">
    $21,461
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2017
   </td>
   <td style="text-align:center">
    $11,759
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2016
   </td>
   <td style="text-align:center">
    $7,000
   </td>
  </tr>
  <tr>
   <td style

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [151]:
pd.read_html(str(tables[0]), flavor='bs4')

[    Tesla Annual Revenue (Millions of US $)  \
 0                                      2021   
 1                                      2020   
 2                                      2019   
 3                                      2018   
 4                                      2017   
 5                                      2016   
 6                                      2015   
 7                                      2014   
 8                                      2013   
 9                                      2012   
 10                                     2011   
 11                                     2010   
 12                                     2009   
 
    Tesla Annual Revenue (Millions of US $).1  
 0                                    $53,823  
 1                                    $31,536  
 2                                    $24,578  
 3                                    $21,461  
 4                                    $11,759  
 5                                    

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [216]:
tesla_revenue["Revenue"] =  pd.to_numeric(tesla_revenue['Revenue'].replace('[\$,]', '', regex=True), errors='coerce')


          Date  Revenue
0   2022-09-30    21454
1   2022-06-30    16934
2   2022-03-31    18756
3   2021-12-31    17719
4   2021-09-30    13757
5   2021-06-30    11958
6   2021-03-31    10389
7   2020-12-31    10744
8   2020-09-30     8771
9   2020-06-30     6036
10  2020-03-31     5985
11  2019-12-31     7384
12  2019-09-30     6303
13  2019-06-30     6350
14  2019-03-31     4541
15  2018-12-31     7226
16  2018-09-30     6824
17  2018-06-30     4002
18  2018-03-31     3409
19  2017-12-31     3288
20  2017-09-30     2985
21  2017-06-30     2790
22  2017-03-31     2696
23  2016-12-31     2285
24  2016-09-30     2298
25  2016-06-30     1270
26  2016-03-31     1147
27  2015-12-31     1214
28  2015-09-30      937
29  2015-06-30      955
30  2015-03-31      940
31  2014-12-31      957
32  2014-09-30      852
33  2014-06-30      769
34  2014-03-31      621
35  2013-12-31      615
36  2013-09-30      431
37  2013-06-30      405
38  2013-03-31      562
39  2012-12-31      306
40  2012-09-30  

Execute the following lines to remove an null or empty strings in the Revenue column.


In [218]:
tesla_revenue.dropna(inplace=True)

tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]


Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [219]:
tesla_revenue.tail()

Date  Revenue
48  2010-09-30       31
49  2010-06-30       28
50  2010-03-31       21
52  2009-09-30       46
53  2009-06-30       27

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [160]:
GameStop = yf.Ticker("GME")

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [161]:
GameStop = GameStop.history(period="max")


**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [163]:
GameStop.reset_index(inplace=True)
GameStop.head()

Date      Open      High       Low     Close    Volume  \
0 2002-02-13 00:00:00-05:00  1.620128  1.693350  1.603296  1.691666  76216000   
1 2002-02-14 00:00:00-05:00  1.712707  1.716074  1.670626  1.683250  11021600   
2 2002-02-15 00:00:00-05:00  1.683250  1.687458  1.658002  1.674834   8389600   
3 2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578047  1.607504   7410400   
4 2002-02-20 00:00:00-05:00  1.615920  1.662210  1.603296  1.662210   6892800   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [164]:
url ="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html"
html_data=requests.get(url).text

Parse the html data using `beautiful_soup`.


In [169]:
soup = BeautifulSoup(html_data, "html.parser")
for index,table in enumerate(tables):
    if ("GameStop Revenue" in str(table)):
        table_index = index
print(table_index)
print(tables[table_index].prettify())


0
<table class="historical_data_table table">
 <thead>
  <tr>
   <th colspan="2" style="text-align:center">
    Tesla Annual Revenue
    <br/>
    <span style="font-size:14px;">
     (Millions of US $)
    </span>
   </th>
  </tr>
 </thead>
 <tbody>
  <tr>
   <td style="text-align:center">
    2021
   </td>
   <td style="text-align:center">
    $53,823
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2020
   </td>
   <td style="text-align:center">
    $31,536
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2019
   </td>
   <td style="text-align:center">
    $24,578
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2018
   </td>
   <td style="text-align:center">
    $21,461
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2017
   </td>
   <td style="text-align:center">
    $11,759
   </td>
  </tr>
  <tr>
   <td style="text-align:center">
    2016
   </td>
   <td style="text-align:center">
    $7,000
   </td>
  </tr>
  <tr>
   <td style

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [236]:
gme_revenue = pd.read_html(str(tables[1]), flavor='bs4')[0]
gme_revenue.columns=['Date','Revenue']
#print(gme_revenue)
gme_revenue["Revenue"] =  pd.to_numeric(gme_revenue['Revenue'].replace('[\$,]', '', regex=True), errors='coerce')

gme_revenue.dropna(inplace=True)
gme_revenue = gme_revenue[gme_revenue['Revenue'] != ""]
print(gme_revenue)

          Date  Revenue
0   2022-09-30  21454.0
1   2022-06-30  16934.0
2   2022-03-31  18756.0
3   2021-12-31  17719.0
4   2021-09-30  13757.0
5   2021-06-30  11958.0
6   2021-03-31  10389.0
7   2020-12-31  10744.0
8   2020-09-30   8771.0
9   2020-06-30   6036.0
10  2020-03-31   5985.0
11  2019-12-31   7384.0
12  2019-09-30   6303.0
13  2019-06-30   6350.0
14  2019-03-31   4541.0
15  2018-12-31   7226.0
16  2018-09-30   6824.0
17  2018-06-30   4002.0
18  2018-03-31   3409.0
19  2017-12-31   3288.0
20  2017-09-30   2985.0
21  2017-06-30   2790.0
22  2017-03-31   2696.0
23  2016-12-31   2285.0
24  2016-09-30   2298.0
25  2016-06-30   1270.0
26  2016-03-31   1147.0
27  2015-12-31   1214.0
28  2015-09-30    937.0
29  2015-06-30    955.0
30  2015-03-31    940.0
31  2014-12-31    957.0
32  2014-09-30    852.0
33  2014-06-30    769.0
34  2014-03-31    621.0
35  2013-12-31    615.0
36  2013-09-30    431.0
37  2013-06-30    405.0
38  2013-03-31    562.0
39  2012-12-31    306.0
40  2012-09-30  

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [214]:
gme_revenue.tail()

Date  Revenue
48  2010-09-30     31.0
49  2010-06-30     28.0
50  2010-03-31     21.0
52  2009-09-30     46.0
53  2009-06-30     27.0

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [ ]:

make_graph(tesla_data, tesla_revenue, 'Tesla')

Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [ ]:
make_graph(GameStop,gme_revenue,'GameStop')

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
